go-controller/pkg/allocator/mac/reservation.go

```diff
@@ -0,0 +1,61 @@
+package mac
+
+import (
+	"errors"
+	"net"
+	"sync"
+)
+
+// ReservationManager tracks reserved MAC addresses requests of pods and detect MAC conflicts,
+// where one pod request static MAC address that is used by another pod.
+type ReservationManager struct {
+	// lock for storing a MAC reservation.
+	lock sync.Mutex
+	// store for reserved MAC address request by owner. Key is MAC address, value is owner identifier.
+	store map[string]string
+}
+
+// NewManager creates a new ReservationManager.
+func NewManager() *ReservationManager {
+	return &ReservationManager{
+		store: make(map[string]string),
+	}
+}
+
+var ErrMACConflict = errors.New("MAC address already in use")
+var ErrMACReserved = errors.New("MAC address already reserved for the given owner")
```

Comment:
so this error will not trigger a rollback, right ?

Reply:
Right, since pod-allocation can run multiple times per pod event, it should ignore this error to prevent MAC begin released on rollback while still being used by legitimate pod.

For example, on VM migration where the migration destination pod is running, pod-allocation should be ignore repeated  requests (same pod, same mac and owner) followed by ErrMACReserved. 
This change is the missing piece to keep  pod-alloation idempotent